src/schema.rs

```diff
@@ -0,0 +1,462 @@
+        let fields = self
+            .children
+            .iter()
+            .map(|(name, dt)| {
+                let dt = dt.to_arrow_data_type();
+                Field::new(name, dt, true)
+            })
+            .collect::<Vec<_>>();
+        Schema::new_with_metadata(fields, user_metadata.clone())
+    }
+
+    /// Project only specific columns from the root type by column name.
+    pub fn project<T: AsRef<str>>(&self, fields: &[T]) -> Self {
+        // TODO: change project to accept project mask (vec of bools) instead of relying on col names?
+        // TODO: be able to nest project? (i.e. project child struct data type) unsure if actually desirable
+        let fields = fields.iter().map(AsRef::as_ref).collect::<Vec<_>>();
+        let children = self
+            .children
+            .iter()
+            .filter(|c| fields.contains(&c.0.as_str()))
+            .map(|c| c.to_owned())
+            .collect::<Vec<_>>();
+        Self { children }
+    }
+
+    /// Construct from protobuf types.
+    pub fn from_proto(types: &[proto::Type]) -> Result<Self> {
+        ensure!(!types.is_empty(), NoTypesSnafu {});
+        let children = parse_struct_children_from_proto(types, 0)?;
+        Ok(Self { children })
+    }
+}
+
+impl Display for RootDataType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ROOT")?;
+        for child in &self.children {
+            write!(f, "\n  {} {}", child.0, child.1)?;
+        }
+        Ok(())
+    }
+}
+
+/// Helper function since this is duplicated for [`RootDataType`] and [`DataType::Struct`]
+/// parsing from proto.
+fn parse_struct_children_from_proto(
+    types: &[proto::Type],
+    column_index: usize,
+) -> Result<Vec<(String, DataType)>> {
+    // These pre-conditions should always be upheld, especially as this is a private function
+    assert!(column_index < types.len());
+    let ty = &types[column_index];
+    assert!(ty.kind() == proto::r#type::Kind::Struct);
+    ensure!(
+        ty.subtypes.len() == ty.field_names.len(),
+        UnexpectedSnafu {
+            msg: format!(
+                "Struct type for column index {} must have matching lengths for subtypes and field names lists",
+                column_index,
+            )
+        }
+    );
+    let children = ty
+        .subtypes
+        .iter()
+        .zip(ty.field_names.iter())
+        .map(|(&index, name)| {
+            let index = index as usize;
+            let name = name.to_owned();
+            let dt = DataType::from_proto(types, index)?;
+            Ok((name, dt))
+        })
+        .collect::<Result<Vec<_>>>()?;
+    Ok(children)
+}
+
+/// Represents the exact data types supported by ORC.
+///
+/// Each variant holds the column index in order to associate the type
+/// with the specific column data present in the stripes.
+#[derive(Debug, Clone)]
+pub enum DataType {
+    /// 1 bit packed data.
+    Boolean { column_index: usize },
```

Comment:
Not sure about keeping the column index here, as it ties the data type to a specific column instead of representing a type by itself, but is useful in terms of compound types (as you can get all the child column indexes easily)